streamx-plugin/streamx-flink-packer/src/main/scala/com/streamxhub/streamx/flink/packer/pipeline/DockerResolveProgress.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2019 The StreamX Project
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.streamxhub.streamx.flink.packer.pipeline
+
+import com.github.dockerjava.api.model.{PullResponseItem, PushResponseItem}
+import com.streamxhub.streamx.flink.packer.pipeline.BuildPipelineHelper.calPercent
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * cache storage for docker resolved progress
+ *
+ * @author Al-assad
+ */
+class DockerResolveProgress(val pull: DockerPullProgress, val build: DockerBuildProgress, val push: DockerPushProgress)
+
+class DockerPullProgress(val layers: mutable.Map[String, DockerLayerProgress], var error: String, var lastTime: Long) {
+  //noinspection DuplicatedCode
+  def update(pullRsp: PullResponseItem): Unit = {
+    if (pullRsp == null || pullRsp.getId == null || pullRsp.getStatus == null) {
+      return
+    }
+    if (pullRsp.getStatus.contains("complete")) {
+      layers += pullRsp.getId -> DockerLayerProgress(pullRsp.getId, pullRsp.getStatus, 1, 1)
+      lastTime = System.currentTimeMillis
+    } else {
+      val cur =
```

Comment:
Try(pullRsp.getProgressDetail.getCurrent.toLong).getOrElse(0)

Would this be better?

Reply:
Relying on exception catching as conditional control is inefficient in execution, especially in this high-frequency computing scenario.